### PR TITLE
Building on the TD Shellmap, plus some more lua bindings

### PR DESCRIPTION
**Pull request by Nelax**
I'm rewriting the lua script for the TD shellmap given it's lack of actual activity. It still has quite a bit to go as you can see but it does far more than the current script already.

Update: Fixed all noted errors and added the actual production scripting to make it a more active shellmap. I'm not really concerned with adding more activity to the map but it could use a couple things like having an orca group fly in for the helipads, maybe adding an A-10 airstrike somewhere, or having Nod fly in an infantry group across the eastern river to attack across the bridge.

### CODE REVIEW

File: OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs [hunk 1]
```diff
@@ -54,5 +54,12 @@ public void Scatter()
 		{
 			Self.Trait<Mobile>().Nudge(Self, Self, true);
 		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to and enter a transport.")]
+		public void GetInTransport(Actor transport)
+		{
+				Self.QueueActivity(new EnterTransport(Self, transport, 1, true));
```
Comment (penev92) [inline OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs]: Single tab indentation.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs [hunk 2]
```diff
@@ -73,5 +73,17 @@ public void PatrolUntil(CPos[] waypoints, LuaFunction func, int wait = 0)
 				using (var f = func.CopyReference() as LuaFunction)
 					Self.QueueActivity(new CallFunc(() => PatrolUntil(waypoints, f, wait)));
 		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to a cell and demolish whatever is there.")]
+		public void GoDemolish(Actor target)
+		{
+			if (Self.HasTrait<C4Demolition>() && !target.IsDead)
```
Comment (penev92) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: Style nit: please invert `if` statement to reduce nesting.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs [hunk 3]
```diff
@@ -73,5 +73,17 @@ public void PatrolUntil(CPos[] waypoints, LuaFunction func, int wait = 0)
...
+			{
+				var info = Self.Info.Traits.GetOrDefault<C4DemolitionInfo>();
+				Self.QueueActivity(new Demolish(Self, target, info.C4Delay, info.Flashes,
```
Comment (penev92) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: The automatic build failed because of the empty space at the end of this line.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs [hunk 4]
```diff
@@ -73,5 +73,17 @@ public void PatrolUntil(CPos[] waypoints, LuaFunction func, int wait = 0)
 				using (var f = func.CopyReference() as LuaFunction)
 					Self.QueueActivity(new CallFunc(() => PatrolUntil(waypoints, f, wait)));
 		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to a cell and demolish whatever is there.")]
+		public void GoDemolish(Actor target)
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: This will need to move into its own `DemolishProperties` that implements `Requires<C4Demolition>` in order for it to behave correctly in all cases.

File: OpenRA.Mods.Common/Scripting/Properties/GeneralProperties.cs [hunk 5]
```diff
@@ -89,6 +89,12 @@ public GeneralProperties(ScriptContext context, Actor self)
 		[Desc("The actor position in cell coordinates.")]
 		public CPos Location { get { return Self.Location; } }
 
+		[Desc("The x value of the actor position.")]
+		public int X { get { return Self.Location.X; } }
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/GeneralProperties.cs]: These aren't necessary - just use Location.X and Location.Y in the script itself.

Comment (Bynnar18) [inline OpenRA.Mods.Common/Scripting/Properties/GeneralProperties.cs]: My bad, I was only using the Lua API as reference and didn't think about trying it.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs (hunk 3 above)
Comment (Bynnar18) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: I'll have that fixed.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs (hunk 2 above)
Comment (Bynnar18) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: Given I've moved it to require C4Demolition, do you mean just something like this:

if (target.IsDead) { return; }
...

File: OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs (hunk 1 above)
Comment (Bynnar18) [inline OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs]: I'll get this fixed.

File: OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs (hunk 2 above)
Comment (penev92) [inline OpenRA.Mods.Common/Scripting/Properties/CombatProperties.cs]: Yes, without the curly brackets `{ }`.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 6]
```diff
@@ -1,3 +1,248 @@
...
+	Trigger.OnIdle(commandos[1], function()
+			if(Map.NamedActor("Turret1") ~= nil) then
+				commandos[1].GoDemolish(Turret1)
+			end
+			if(Map.NamedActor("beachdude1") ~= nil or Map.NamedActor("beachdude2") ~= nil) then
+				commandos[1].AttackMove(CommMove.Location)
+			end
+			if(Map.NamedActor("Turret3") ~= nil) then
+				commandos[1].GoDemolish(Turret3)
+				commandos[1].AttackMove(CommFin.Location)
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: You don't need that here, as you repeat it in [line 65]().

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 7]
```diff
@@ -1,3 +1,248 @@
...
+	Trigger.OnIdle(commandos[2], function()
+			if(Map.NamedActor("Turret2") ~= nil) then
+				commandos[2].GoDemolish(Turret2)
+			end
+			if(Map.NamedActor("beachdude1") ~= nil or Map.NamedActor("beachdude2") ~= nil) then
+				commandos[2].AttackMove(CommMove.Location)
+			end
+			if(Map.NamedActor("Turret4") ~= nil) then
+				commandos[2].GoDemolish(Turret4)
+				commandos[2].AttackMove(CommFin.Location)
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Same here.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 8]
```diff
@@ -1,3 +1,248 @@
...
+
+SendCommandos = function()
+	local commandos = Reinforcements.ReinforceWithTransport(gdi, "tran", CommandoTeam, 
+	{ChinEntry.Location, ChinUnload.Location}, {ChinExit.Location})[2]
+	
+	Trigger.OnIdle(commandos[1], function()
+			if(Map.NamedActor("Turret1") ~= nil) then
+				commandos[1].GoDemolish(Turret1)
+			end
+			if(Map.NamedActor("beachdude1") ~= nil or Map.NamedActor("beachdude2") ~= nil) then
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Why do you check both commandes but only move one?

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 9]
```diff
@@ -1,3 +1,248 @@
...
+	tank.ScriptedMove(LSTLLand.Location - CVec.New(11,7))
+	MakeUnitInvincible(tank)
+end
+
+MCVAction = function(mcv)
+	mcv.ScriptedMove(LSTMLand.Location - CVec.New(0,2))
+	mcv.Wait(200)
+	mcv.ScriptedMove(MCVDeploy.Location)
+	mcv.CallFunc(function() Trigger.AfterDelay(100, function() BuildBase() end) end)
+	Trigger.OnIdle(mcv, function() mcv.Deploy()	end)
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Unintentional tab after ```mcv.Deploy()```.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 10]
```diff
@@ -5,31 +250,17 @@ speed = 5
...
 
-LoopTrack = function(actor, left, right)
-	actor.ScriptedMove(left)
-	actor.Teleport(right)
-	actor.CallFunc(function() LoopTrack(actor, left, right) end)
-end
+	--SkipCommandos()
+	SendCommandos()
 
-PlayMusic = function()
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Why did you remove this?

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 11]
```diff
@@ -1,3 +1,248 @@
...
+				commandos[2].GoDemolish(Turret4)
+				commandos[2].AttackMove(CommFin.Location)
+			end
+			commandos[2].AttackMove(CommFin.Location)
+		 end)
+
+	Trigger.OnKilled(Turret1, function()
+			beachdude1.AttackMove(BeachDudeMove.Location)
+			beachdude2.AttackMove(BeachDudeMove.Location)
+		end)
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: We usually have indention that way:
```
	Trigger.OnKilled(Turret1, function() 
		beachdude1.AttackMove(BeachDudeMove.Location) 
		beachdude2.AttackMove(BeachDudeMove.Location) 
	end) 
```

Comment (penev92) [inline mods/cnc/maps/shellmap/shellmap.lua]: Meaning a single tab. It's often ugly, but it's the rules.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 8 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: I'm not quite sure what you're referring to. I send each commando at two different turrets.

Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Ah sorry, i missunderstood this here.

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+		public DemolitionProperties(ScriptContext context, Actor self)
+			: base(context, self) { }
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to a cell and demolish whatever is there.")]
+		public void GoDemolish(Actor target)
+		{
+			if (target.IsDead)
+				return;
+			var info = Self.Info.Traits.GetOrDefault<C4DemolitionInfo>();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs]: A new line after the ```return;```, please.

File: OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs [hunk 13]
```diff
@@ -54,5 +54,12 @@ public void Scatter()
 		{
 			Self.Trait<Mobile>().Nudge(Self, Self, true);
 		}
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to and enter a transport.")]
+		public void GetInTransport(Actor transport)
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Scripting/Properties/MobileProperties.cs]: I think ```EnterTransport``` would be a better name.

File: OpenRA.Mods.Common/Scripting/Properties/PlayerProperties.cs [hunk 14]
```diff
@@ -32,5 +33,17 @@ public Actor[] GetGroundAttackers()
...
+		[Desc("Gives the player money. Don't use to reduce.")]
+		public void GiveMoney(int amount)
+		{
+			Player.PlayerActor.Trait<PlayerResources>().GiveCash(amount);
+		}
+
+		[Desc("Takes money from the player.")]
+		public void TakeMoney(int amount)
+		{
+			Player.PlayerActor.Trait<PlayerResources>().TakeCash(amount);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Scripting/Properties/PlayerProperties.cs]: This should be moved to https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Scripting/Properties/ResourceProperties.cs and make use of the cached `PlayerResources` trait there.

Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/PlayerProperties.cs]: ResourceProperties already has a `Cash` property for adding/removing cash, so these are entirely unnecessary.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 15]
```diff
@@ -1,3 +1,358 @@
...
+	end)
+	clock.CallFunc(function() Trigger.AfterDelay(Actor.BuildTime("gtwr"), function()
+			MakeUnitInvincible(Reinforcements.Reinforce(gdi, { "gtwr" }, { (MCVDeploy.Location+CVec.New(-1,-6)) })[1])
+		end)
+	end)
+
+	--Build a Refinery and make harvester immune, then build another Guard Tower
+	clock.Wait(Actor.BuildTime("proc"))
+	clock.CallFunc(function()
+		Reinforcements.Reinforce(gdi, { "proc" }, { (MCVDeploy.Location+CVec.New(4,-4)) })
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Please use ```Actor.Create("proc", true, { Owner = gdi, Location = MCVDeploy.Location + CVec.New(4, -4) })``` instead
(here and for the creating of the other buildings).

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 16]
```diff
@@ -5,31 +360,16 @@ speed = 5
...
-end
 
-LoopTrack = function(actor, left, right)
-	actor.ScriptedMove(left)
-	actor.Teleport(right)
-	actor.CallFunc(function() LoopTrack(actor, left, right) end)
-end
+	SendCommandos()
 
-PlayMusic = function()
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Can you revert this, please? It's for playing the menu song.

Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: I did at some point and put the function near the top. If it isn't there now, it'll be in next commit for sure.

**Comment by penev92** [comment]
Leave a comment when updating, because GitHub doesn't send notifications when you push. ;)

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+{
+	[ScriptPropertyGroup("Demolition")]
+	public class DemolitionProperties : ScriptActorProperties, Requires<IMoveInfo>, Requires<C4DemolitionInfo>
+	{
+		public DemolitionProperties(ScriptContext context, Actor self)
+			: base(context, self) { }
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move to a cell and demolish whatever is there.")]
+		public void GoDemolish(Actor target)
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs]: I think it would be more consistent to call this `Demolish`, with a description "Move to and demolish an actor".

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+		{
+			if (target.IsDead)
+				return;
+
+			var info = Self.Info.Traits.GetOrDefault<C4DemolitionInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs]: As a general rule, if you use GetOrDefault (or TraitOrDefault) then you should always do a null-check before using the result.  It is guaranteed to be non-null here (because of the `Requires<C4DemolitionInfo>`), but you should still change this to `.Get<C4DemolitionInfo>()` for clarity.

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+using Eluant;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Scripting;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Scripting
+{
+	[ScriptPropertyGroup("Demolition")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs]: This probably wants to live in the "Combat" group.

**Comment by Bynnar18** [comment]
Just did another push to include the requested changes.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 20]
```diff
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (obrakmann) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Please try to avoid committing changes like this and the last hunk in this file (last line)

File: OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+			: base(context, self) { }
+
+		[ScriptActorPropertyActivity]
+		[Desc("Move and demolish an actor.")]
+		public void Demolish(Actor target)
+		{
+			if (target.IsDead)
+				return;
+
+			var info = Self.Info.Traits.Get<C4DemolitionInfo>();
```
Comment (obrakmann) [inline OpenRA.Mods.Common/Scripting/Properties/DemolitionProperties.cs]: It would be nicer to do this in the constructor (and store it in a private field).

**Comment by abcdefg30** [comment]
Please squash the "fix commits".

File: mods/cnc/maps/shellmap/map.yaml [hunk 22]
```diff
@@ -182,7 +180,7 @@ Actors:
 	Actor40: t06
 		Location: 8,6
 		Owner: Neutral
-	Actor43: afld
+	NodAfld: afld
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/map.yaml]: We've recently agreed on putting all the actors with custom names at the end of the list. It would be nice if you could do that here, too.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 23]
```diff
@@ -1,3 +1,362 @@
+CommandoTeam = {"rmbo", "rmbo"}
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: Please use spaces before closing and after opening braces.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 24]
```diff
@@ -1,3 +1,362 @@
...
+end
+--Repeated Functions END
+
+--Event Sequence Functions START
+SendCommandos = function()
+	local commandos = Reinforcements.ReinforceWithTransport(gdi, "tran", CommandoTeam, 
+	{ChinEntry.Location, ChinUnload.Location}, {ChinExit.Location})[2]
+	
+	Trigger.OnIdle(commandos[1], function()
+			if(Map.NamedActor("Turret1") ~= nil) then
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: I think using `if not Turret1.IsDead then` works better here.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 25]
```diff
@@ -1,3 +1,362 @@
...
+				commandos[2].EnterTransport(trans)
+			end
+		end)
+	end)
+end
+
+Beachhead = function()
+	--todo: Make custom reinforce function to handle LST Dropoff properly
+	local ltank = Reinforcements.ReinforceWithTransport(gdi, "lst", {"htnk"}, LSTLEntryPath, LSTLExitPath, UnloadLST)[2]
+	Trigger.OnAddedToWorld(ltank[1], function(unit) LeftTankAction(unit) end)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: You can just use the `actionFunc` parameter in `ReinforceWithTransport` for that.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 26]
```diff
@@ -1,3 +1,362 @@
...
+			commandos[2].Stop()
+			commandos[2].ScriptedMove(CommFin.Location)
+			Trigger.Clear(commandos[2], "OnIdle")
+	end)
+	Trigger.OnAllKilled({beachdude1, beachdude2}, function()
+		local trans = Reinforcements.ReinforceWithTransport(gdi, "tran", nil, {EvacEntry.Location, EvacLand.Location})[1]
+		Trigger.OnIdle(trans, function()
+			if(trans.HasPassengers) then
+				trans.Wait(25)
+				trans.CallFunc(function() PlayMusic() end)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: You already start the music in the `WorldLoaded` function, and here again. Is that on purpose?

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 27]
```diff
@@ -1,3 +1,362 @@
...
+	end
+end
+--Repeated Functions END
+
+--Event Sequence Functions START
+SendCommandos = function()
+	local commandos = Reinforcements.ReinforceWithTransport(gdi, "tran", CommandoTeam, 
+	{ChinEntry.Location, ChinUnload.Location}, {ChinExit.Location})[2]
+	
+	Trigger.OnIdle(commandos[1], function()
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: I'm a bit confused why you're using the `OnIdle` trigger so much. (Almost?) All the things you make your actors do are queued actions, so you can just put all the calls one behind the other and it'll work fine.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 28]
```diff
@@ -1,3 +1,362 @@
...
+LSTREntryPath = {LSTRStart.Location, LSTRMove.Location, LSTRLand.Location}
+LSTMEntryPath = {LSTMStart.Location, LSTMMove.Location, LSTMLand.Location}
+LSTLEntryPath = {LSTLStart.Location, LSTLMove.Location, LSTLLand.Location}
+
+LSTRExitPath = {CPos.New(LSTRMove.Location.X,LSTLStart.Location.Y), CPos.New(8,LSTLStart.Location.Y)}
+LSTMExitPath = {LSTMMove.Location, CPos.New(8,LSTMStart.Location.Y)}
+LSTLExitPath = {CPos.New(LSTLMove.Location.X,LSTRStart.Location.Y), CPos.New(8,LSTRStart.Location.Y)}
+
+--Repeated Functions START
+UnloadLST = function(lst,cargo)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: The `UnloadPassengers` and `Reinforcements.ReinforceWithTransports` functions do this automatically for you

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 29]
```diff
@@ -1,3 +1,362 @@
...
+	Media.PlayMusic("map1", PlayMusic)
+end
+
+StayInLocation = function(a,loc,dist)
+	if not a.IsDead then
+		local vector = CVec.New(a.Location.X-loc.X,a.Location.Y-loc.Y)
+		if math.abs(vector.X) > dist or math.abs(vector.Y) > dist then
+			a.Stop()
+			a.AttackMove(loc)
+			Trigger.AfterDelay(250, function() StayInLocation(a,loc,dist) end)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: Please use the `DateTime.Seconds` and/or `DateTime.Minutes` functions instead of hard-coded numbers for better readability.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 30]
```diff
@@ -1,3 +1,362 @@
...
+		end
+	end
+end
+MakeNod = function(fact,teams)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: Newline between functions, please

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 31]
```diff
@@ -1,3 +1,362 @@
...
+
+LSTREntryPath = {LSTRStart.Location, LSTRMove.Location, LSTRLand.Location}
+LSTMEntryPath = {LSTMStart.Location, LSTMMove.Location, LSTMLand.Location}
+LSTLEntryPath = {LSTLStart.Location, LSTLMove.Location, LSTLLand.Location}
+
+LSTRExitPath = {CPos.New(LSTRMove.Location.X,LSTLStart.Location.Y), CPos.New(8,LSTLStart.Location.Y)}
+LSTMExitPath = {LSTMMove.Location, CPos.New(8,LSTMStart.Location.Y)}
+LSTLExitPath = {CPos.New(LSTLMove.Location.X,LSTRStart.Location.Y), CPos.New(8,LSTRStart.Location.Y)}
+
+--Repeated Functions START
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: These types of comments are unnecessary.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 32]
```diff
@@ -1,3 +1,362 @@
...
+	tank.ScriptedMove(LSTLLand.Location - CVec.New(11,5))
+	MakeUnitInvincible(tank)
+end
+
+MCVAction = function(mcv)
+	--mcv.ScriptedMove(LSTMLand.Location - CVec.New(0,1))
+	mcv.Wait(250)
+	mcv.ScriptedMove(MCVDeploy.Location)
+	mcv.CallFunc(function() Trigger.AfterDelay(100, function() BuildBase() end) end)
+	Trigger.OnIdle(mcv, function() mcv.Deploy()	end)
```
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: This trigger is unnecessary as well. Just call `mcv.Deploy()`, the action will be queued.

**Comment by obrakmann** [comment]
I like those improvements to the shellmap, equally interesting as the ra shellmap to watch and much better than the previous quiet one.

However, it shows a bit that you're probably not familiar with our Lua API yet. You're using triggers a bit too much and don't use the existing functions to their full potential.

I suspect the reason may be that you didn't properly pick up on the fact that many of the actions you can make actors do are not actually executed right when you call them, but instead are put on a queue of actions that the actors then perform one after the other, asynchronously to what's happening in the lua script. This takes a while to get used to.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 28 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: This was necessary to get the proper exit direction out of the units. Otherwise, they exit Northeast unlike they should.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 25 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: That comment was outdated, UnloadLST() did just that.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 33]
```diff
@@ -1,3 +1,362 @@
...
+	end)
+end
+
+Beachhead = function()
+	--todo: Make custom reinforce function to handle LST Dropoff properly
+	local ltank = Reinforcements.ReinforceWithTransport(gdi, "lst", {"htnk"}, LSTLEntryPath, LSTLExitPath, UnloadLST)[2]
+	Trigger.OnAddedToWorld(ltank[1], function(unit) LeftTankAction(unit) end)
+	Trigger.AfterDelay(55, function()
+		local mcv = Reinforcements.ReinforceWithTransport(gdi, "lst", {"mcv"}, LSTMEntryPath, LSTMExitPath, UnloadLST)[2]
+		Trigger.OnAddedToWorld(mcv[1], function(unit) MCVAction(unit) end)
```
Comment (ScottNZ) [inline mods/cnc/maps/shellmap/shellmap.lua]: This can just be `Trigger.OnAddedToWorld(mcv[1], MCVAction)` (and same for the other trigger setups)

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 27 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: Removing the OnIdle caused the commandos to do absolutely nothing. Uising an OnAddedToWorld might work equally well but OnIdle means they are given something to do if they somehow get lost.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 26 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: Yes, this calls a different function. The initial call plays the music that was on the TD score screen, I found it to be quite fitting for the commando sequence. This one calls the script's own lua function to loop the standard music after the initial song is finished which is at this point.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 33 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: I'll clean up a bit with this information. I came to the conclusion that I needed a new function since I was calling it with parentheses and getting nothing but errors.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 20 above)
Comment (Bynnar18) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: I'm not quite sure why such a thing happens in the first place.

**Comment by Bynnar18** [comment]
I would not of tried something as complex as this if it weren't for the fact that no one seems to have been giving Tiberian Dawn any attention and I saw that release after release TD kept the same old boring incomplete shellmap. Clearly, this is my first time creating anything with Lua and I'm willing to listen to advice and fix things as they come up. 

While I realize that I may have used a large amount of triggers, I did so out of trial and error. I found that many times a unit would disregard a directive given, that is the primary reason that I have as many triggers as I do. Aside from that, this shellmap isn't as passive as the Red Alert one, it has alot more moving pieces that lead up to the loop of units fighting.

File: OpenRA.Test/OpenRA.Test.csproj [hunk 34]
```diff
@@ -14,26 +14,26 @@
   </PropertyGroup>
   <PropertyGroup Condition="'$(Configuration)|$(Platform)' == 'Debug|x86'">
     <DebugSymbols>true</DebugSymbols>
-    <OutputPath>..\</OutputPath>
+    <OutputPath>.</OutputPath>
```
Comment (Mailaender) [inline OpenRA.Test/OpenRA.Test.csproj]: This broke the Windows build. https://ci.appveyor.com/project/OpenRA/openra/build/1.0.1028

Please don't commit any of these unrelated changes to the project files.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 20 above)
Comment (obrakmann) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: Because you're probably using VS and it insists on adding a UTF-8 BOM to files it touches.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 27 above)
Comment (obrakmann) [inline mods/cnc/maps/shellmap/shellmap.lua]: That's strange. It should just work when you use the `actionFunc` argument in the `ReinforceWithTransport` call. Or did you do it differently?

File: OpenRA.Test/OpenRA.Test.csproj (hunk 34 above)
Comment (Bynnar18) [inline OpenRA.Test/OpenRA.Test.csproj]: My bad, that must've happened when I was squashing commits. Monodevelop has a tendency to make stupid changes like this for no reason. I'll get it fixed as soon as I can.

**Comment by Bynnar18** [comment]
Ok, I think I got it. I could've made a mess though. I went though a half dozen resets before I figured out how to use git revert properly.

**Comment by Mailaender** [comment]
https://help.github.com/articles/about-git-rebase/ may be the command you are looking for to clean up all those junk commits.

File: mods/cnc/maps/shellmap/shellmap.lua (hunk 27 above)
Comment (Bynnar18) [inline mods/cnc/maps/shellmap/shellmap.lua]: I gave that a quick shot, I couldn't really get the heli to drop off the commandos properly. I probably missed something though. Regardless, this way works just fine for now.

**Comment by Bynnar18** [comment]
I already reverted the commit and rebase claims it's up to date.

**Comment by Mailaender** [comment]
See http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html on how to get rid of the fixup and revert commits.

**Comment by Bynnar18** [comment]
Well like I said, I already reverted it. Anyhow, I squashed a bunch of the commits. I don't intend to do anything else with the lua I don't think since whenever I've tried changing things to use less triggers I end up breaking the map in one way or another.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 35]
```diff
@@ -1,35 +1,370 @@
...
+		Trigger.OnAllKilled(team,function()
+			if not fact.IsDead then
+				MakeGDI(fact,teamtype)
+			end
+		end)
+	end)
+end
+
+MakeUnitInvincible = function(unit)
+	if not unit.HasProperty("AcceptsUpgrade") then
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: Just
```
	if unit.AcceptsUpgrade("unkillable") then
		unit.GrantUpgrade("unkillable")
		unit.Stance = "Defend"
	end
```
should work as well.

File: mods/cnc/maps/shellmap/shellmap.lua [hunk 36]
```diff
@@ -1,35 +1,370 @@
...
+GDIVehicleTeam = { "jeep", "jeep" }
+NodInfantry =
+	{
+		{{"e1", "e1", "e3"},function(a)
+			a.AttackMove(WestSide.Location)
+			a.AttackMove(MCVDeploy.Location)
+		end},
+		{{"e1", "e1", "e1", "e3", "e3"},function(a)
+			a.AttackMove(WestSide.Location)
+			a.AttackMove(MCVDeploy.Location)
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/shellmap.lua]: This adds a lot of duplication.
What about using a function like
```
NodInfantryMove = function(a)
	a.AttackMove(WestSide.Location)
	a.AttackMove(MCVDeploy.Location)
end
```

File: mods/cnc/maps/shellmap/map.yaml [hunk 37]
```diff
@@ -994,18 +1025,102 @@ Rules:
...
+			Speed: 82
 	BOAT:
 		Mobile:
 			Speed: 42
+	MTNK:
+		Buildable:
+			Prerequisites: anyhq
+	LTNK:
+		Buildable:
+			-Prerequisites:
```
Comment (abcdefg30) [inline mods/cnc/maps/shellmap/map.yaml]: I'm not sure, but do we need this?
I mean, you say what the player builds with lua.

Comment (Bynnar18) [inline mods/cnc/maps/shellmap/map.yaml]: The player won't build the units if it doesn't have the prereqs for it, in this case, I'm mostly removing tech level requirements.

Comment (abcdefg30) [inline mods/cnc/maps/shellmap/map.yaml]: Ah, sorry again, I thought you somehow wanted to disable this stuff.

**Comment by abcdefg30** [comment]
Btw, the convention is using ```{ foo }``` and not ```{foo}```.

**Comment by Bynnar18** [comment]
Ok, I went ahead and made independent functions for the Nod attacks but I couldn't get them to work in the array properly without defining a new function there. (Neither { ... , WestAttack } nor { ... , WestAttack(a) } would work)

I'm not sure how much more I'm going to make changes to this. It took me quite a while to get it to work and I'm not really interested in reworking half the script.

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj (hunk 20 above)
Comment (abcdefg30) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: This still needs to be fixed.

**Comment by Bynnar18** [comment]
I'm going to go ahead and close this pull request for now. I'm not really interested in working on this any more right now after putting in all the time in the first place. I won't be getting rid of my branch on github if anyone wants to fix it and I'll probably come back and fix it in the future, but for now it's on hold.

**Comment by pchote** [comment]
We ended up dropping this version of the TD shellmap, so removing from the Future milestone.
